### PR TITLE
Implement basic functionality

### PR DESCRIPTION
**Pull request by tortelio**

### CODE REVIEW

File: src/Dumper.jl [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    mime = get(arguments, :mime, :($STATE.mime))
+    mode = get(arguments, :mode, :($STATE.mode))
+
+    return quote
+        if $STATE.enabled
+            let _path = $isabsolute ? $path : joinpath($directory, $path)
+                $save(_path, $mime, $variable; $mode)
+            end
+        end
+    end |> esc
```
Comment (dhanak) [inline src/Dumper.jl:81]: It is generally not recommended to escape entire expressions, escape individual interpolations instead.

The added benefit, for example, is that you don't have to interpolate `save`.

**Review by dhanak — CHANGES_REQUESTED** [review]
Generally okay, with some glitches.

File: src/Dumper.jl [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    isabsolute = get(arguments, :isabsolute, :($STATE.isabsolute))
+    path = get(arguments, :path, string(variable))
+    directory = get(arguments, :directory, :($STATE.directory))
+    mime = get(arguments, :mime, :($STATE.mime))
+    mode = get(arguments, :mode, :($STATE.mode))
+
+    return quote
+        if $STATE.enabled
+            let _path = $isabsolute ? $path : joinpath($directory, $path)
```
[the remark this answers is not in the document]
Comment (dhanak) [reply src/Dumper.jl:77]: Something along these lines:

```
return if STATE.enabled
    path = isabsolute ? path : joinpath(directory, path)
    :( save($path, $mime, $(esc(variable)), $mode) )
else
    :nothing
end
```

[the remark this answers is not in the document]
Comment (tortelio) [reply src/Dumper.jl:77]: You are kind of right, but I was intentionally made this "dynamic". If you want to switch on-the-fly than you have to implement like this.

File: src/Dumper.jl [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        end
+    end |> esc
+end
+
+###-----------------------------------------------------------------------------
+### File API
+###-----------------------------------------------------------------------------
+
+function save(io::IO, x)::Nothing
+    save(io, MIME"text/plain", x)
```
[the remark this answers is not in the document]
Comment (tortelio) [reply src/Dumper.jl:89]: You are right.

File: src/Dumper.jl [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+###-----------------------------------------------------------------------------
+### State
+###-----------------------------------------------------------------------------
+
+# TODO maybe use some timestamped folder by default
+@kwdef mutable struct State
+    enabled::Bool = false
+    isabsolute::Bool = false
+    mime::MIME = MIME"text/plain"()
```
[the remark this answers is not in the document]
Comment (tortelio) [reply src/Dumper.jl:31]: You are right.

File: src/Utilities.jl [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+###=============================================================================
+
+using Dates: now, format
+
+###=============================================================================
+### Functions
+###=============================================================================
+
+function dump_directory(path::AbstractString = "")::String
+    return joinpath(path, "dump-" * format(now(), "YYYYmmdd-HHMMSS"))
```
[the remark this answers is not in the document]
Comment (tortelio) [reply src/Utilities.jl:20]: Fine.

File: src/Dumper.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (dhanak) [reply src/Dumper.jl:80]: Ok, I see what you mean. I still don't like this macro implementation, it seems too convoluted, and now it calls the `STATE()` function too many times, just to force dynamic behavior.

I'm thinking the following now: put most part of the dump macro body in a function.

```jl
function dump(name, value;
              isabsolute::Bool = STATE().isabsolute,
              directory::AbstractString = STATE().directory,
              path::AbstractString = name,
              [etc.])
   STATE().enabled || return
   path = isabsolute ? path : joinpath(directory, path)
   save(path, mime, variable; mode)
end
                 
And then just call this function from the macro after the dict conversion.

[the remark this answers is not in the document]
Comment (tortelio) [reply src/Dumper.jl:131]: Finally, we couldn't find the straight-forward best solution.

**Pull request merged**